src/file_pipe_log/pipe.rs

```diff
@@ -428,68 +330,105 @@ impl<F: FileSystem> SinglePipe<F> {
+            let mut files = self.active_files.write();
+            let off = (file_seq - files[0].seq) as usize;
+            if off >= files.len() {
+                return Err(box_err!("Active file cannot be purged"));
+            }
+            let mut tail = files.split_off(off);
+            std::mem::swap(&mut tail, &mut files);
+            (files.len(), tail)
+        };
+        let purged_len = purged_files.len();
+        {
+            let remains_capacity = self.capacity.saturating_sub(len);
+            let (recycled_start, mut recycled_len) = {
+                let files = self.recycled_files.read();
+                files
+                    .front()
+                    .map_or_else(|| (1, 0), |f| (f.seq, files.len()))
             };
-            let path = file_id.build_file_path(&self.dir);
-            #[cfg(feature = "failpoints")]
-            {
-                let remove_skipped = || {
-                    fail::fail_point!("file_pipe_log::remove_file_skipped", |_| true);
-                    false
+            let mut new_recycled = VecDeque::new();
+            // The newly purged files from `self.active_files` should be renamed
+            // to recycled files with LOG_APPEND_RESERVED_SUFFIX suffix, to reduce the
+            // unnecessary recovery timecost when restarting.
+            for f in purged_files {
+                let file_id = FileId {
+                    seq: f.seq,
+                    queue: self.queue,
                 };
-                if remove_skipped() {
-                    continue;
+                let path = file_id.build_file_path(&self.paths[f.path_id]);
+                if f.format.version.has_log_signing() && recycled_len < remains_capacity {
+                    let recycled_seq = recycled_start + recycled_len as u64;
+                    let dst_path =
+                        self.paths[f.path_id].join(build_recycled_file_name(recycled_seq));
+                    match self.file_system.reuse_and_open(&path, &dst_path) {
+                        Ok(handle) => {
+                            new_recycled.push_back(File {
+                                seq: recycled_seq,
+                                handle: handle.into(),
+                                path_id: f.path_id,
+                                format: f.format,
+                            });
+                            recycled_len += 1;
+                            continue;
+                        }
+                        Err(e) => error!("failed to reuse purged file {:?}", e),
+                    }
                 }
+                self.file_system.delete(&path)?;
+            }
+            // If there exists several recycled files out of space, contained in
+            // `self.recycled_files` and `purged_files`, we should check and remove them
+            // to avoid the `size` of whole files beyond `self.capacity`.
+            if recycled_len > remains_capacity {
```

Comment:
This block is not covered.

Reply:
It's a stale and redundant judgement here, as we have ensured the remained count of recycled files when restarting.

Reply:
Add a debug assertion.

Reply:
Ping

Reply:
Ping.

Reply:
Supplemented.